### PR TITLE
Make calling Python tools from make more consistent, platform-independent and user-configurable

### PR DESCRIPTION
**Pull request by SpotlightKid**
Rationale:

* When calling Python build tool scripts from makefiles, we should be consistent on how to call them. IMHO it is better to call them via `python </path/to/script>` directly, instead of relying on the correct she-bang line in the script [1] and the executable bit on the script being set. This is more platform-independent.
* The name/path of the Python executable should always be used via the makefile variable `PYTHON` (set in `py/mkenv.mk`. This way it can be easily overwritten by the user (i.e. with `make PYTHON=/path/to/my/python`), which currently doesn't work for the `make-frozen.py` tool as it should (see Damien's [comment here](https://github.com/micropython/micropython/issues/1616#issuecomment-156456904))
* The Python executable name should be part of the value of the makefile variable, which stands for the build tool command (e.g. `MAKE_FROZEN` and `MPY_TOOL`), not part of the command line where it is used. If a Python tool is substituted by another (non-python) program, no change to the Makefiles is necessary, except in `py/mkenv.mk`.
* This also solves #3369 and #1616.

[1] There are systems, where even the assumption that `/usr/bin/env` always exists, doesn't hold true, for example on Android (where otherwise the unix port compiles perfectly well).

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

### CODE REVIEW

**Comment by pfalcon** [comment]
Looks good.

**Review by pfalcon — CHANGES_REQUESTED** [review]
Except for the commit message.

**Comment by pfalcon** [comment]
(First line.) And if github description is so detailed, why is the commit body so bare?

**Comment by SpotlightKid** [comment]
Would that be a good first line?

    py/mk{env,rules}.mk: Use $(PYTHON) consistently when calling Python tools.

I didn't want to put the full reasoning for the PR into the commit message, because some of it is phrased subjectively ("IMHO", etc.). I believe these reasons to be valid, but others might raise different opinions here.

Do want me to put the full description into the commit body (minus the "Signed-off" line)? Feel free to edit the commit message if/when merging. I don't really want to guess what a proper commit message is. The hints [here ](https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md) IMHO aren't really helpful when it comes to the actual wording of the description.

**Comment by stinos** [comment]
Actual wording is of lesser importance, important part is the message explains *why* not just how (after all the latter can be seen in the diff already). The key is basically in the "To get good practical examples of good commits and their messages, browse the git log of the project." line I guess. I remember at one point this was discussed and dpgeorge actually compiled a list of links to what are considered the really good messages but unfortunately I cannot find that back.

**Comment by pfalcon** [comment]
> The hints here IMHO aren't really helpful when it comes to the actual wording of the description.

You're welcome to propose improvements, but first I'd recommend to think better whether it's helpful or not.

There're additional guidelines at https://github.com/micropython/micropython/wiki/ContributorGuidelines#how-to-contribute , in particular, it has:

> Good, detailed commit messages are an easy way to achieve this, and yet sometimes it's better to put informal points, colloquial arguments and big examples in comments to a pull request instead, to keep commit messages concise and focused.

IMHO, that's vice-versa, pretty much answers your questions like "Do want me to put the full description into the commit body?". The answer is "no", it's just strange that a contributor took effort to write 4 paragraph in the PR description (which will be effectively lost after merge) and put almost nothing into the actual commit message, the split should be a bit more balanced.

And the actual commit messages (git log) is almost an infinite source of good commit message examples (ok, maybe 90% of them).

> Feel free to edit the commit message if/when merging.

Well, the whole idea is that contributors, especially returning ones, can deal with that themselves, and a maintainer is not a grunt which cleans up after everybody. That's how it works for any other project out there.

**Review by pfalcon — APPROVED** [review]

**Comment by SpotlightKid** [comment]
> it's just strange that a contributor took effort to write 4 paragraph in the PR description [...]) and put almost nothing into the actual commit message, the split should be a bit more balanced.

You could have just said that and give a hint how the commit message should be.

To be honest, if every time I submit a pull request, I have to put up with this hostile and condescending attitude, this will be my last for this project. I already spent several hours in an effort to make this change easily mergeable, but all I get is complaining. I would appreciate a bit more friendliness and apprehensiveness towards contributions.

**Comment by pfalcon** [comment]
And that's the typical gratefulness - there're guidelines written clearly for everyone, this guy complains they're not helpful, he's being explained what's written in them "on fingers", and here it goes. Have a nice day.

**Comment by SpotlightKid** [comment]
Sorry, Paul, but your communication just rubs me the wrong way. Maybe its my fault, maybe not. But you are definitely putting me off from contributing to MicroPython. That's a pity, but so be it.

**Comment by pfalcon** [comment]
I won't even say in what way your communication rubs me.

**Comment by robert-hh** [comment]
Please calm down.

**Comment by stinos** [comment]
> I won't even say in what way your communication rubs me.

Come on now, you're one of the maintainers, such attitude doesn't help anyone.

**Comment by pfalcon** [comment]
And in all fairness, I find it very non-appreciative to receive comments like:

> I don't really want to guess what a proper commit message is. The hints here https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md IMHO aren't really helpful when it comes to the actual wording of the description.

And when the pretty basic ideas behind it are being explained, as the second round, receive this:

> hostile and condescending attitude

It's really hard to communicate with people who don't put want to slightest effort into learning the conventions, and any attempt at explanation or getting deeper to a problem take as an attack on their skills, effort, or whatever.

**Comment by dpgeorge** [comment]
@SpotlightKid thanks for the contribution and sorry for the fuss.  It's been merged in 1871a924c97cec16d0670d136e9f7056f99865df

**Comment by SpotlightKid** [comment]
Thanks!

**Pull request closed without merging**
